src/System.Slices/System/Buffers/Memory.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a Span<byte> factory</byte>
+    /// </summary>
+    /// <remarks>
+    /// This struct is not safe for multithreaded access, i.e. it's subject to struct tearing that can result in unsafe memory access. 
+    /// </remarks>
+    public struct Memory<T>
+    {
+        public static unsafe Memory<T> Empty = new Memory<T>(null, 0, 0);
+
+        private readonly T[] _array;
+        private readonly int _offset;
+        private readonly unsafe void* _memory;
+        private readonly int _memoryLength;
+
+        public unsafe Memory(void* pointer, int length)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = null;
+            _offset = 0;
+            _memoryLength = length;
+        }
+
+        public unsafe Memory(T[] array) : this(array, 0, array?.Length ?? 0)
+        {
+        }
+
+        public unsafe Memory(T[] array, int offset, int length, bool isPinned = false)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+
+            unsafe
+            {
+                if (isPinned)
+                {
+                    // TODO: Use new Unsafe.AsPointer(ref array[0])
+                    // The caller pinned the array so we can safely get the pointer
+                    _memory = Unsafe.AsPointer(ref array[0]);
```

Comment:
if you don't do the ptr .ctor could do:

`_memory = Unsafe.AsPointer(ref array[offset]);`